### PR TITLE
[FLINK-6783] Changed passing index of type argument while extracting …

### PR DESCRIPTION
**Pull request by dawidwys**
…return type.

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by dawidwys** [comment]
The PR is not fully ready yet, but as it touches critical part, I would like to hear early opinions from somebody more experienced with the Type system. Would be good to know if I am going the right way ;)

**Comment by aljoscha** [comment]
I like the idea of explicitly passing the argument parameters a lot! I would even suggest to remove the old methods with `hasCollector` and so on instead of just deprecating.

Could you please elaborate a bit how the lambda argument arrays work?

**Comment by dawidwys** [comment]
Sure, with the arrays you provide index of the input/output parameters in lambdas. The lambdaInputTypeArgumentIndices[0] is the argument index of lambda function. The lambdaInputTypeArgumentIndices[1] is the index within the argument, if the argument is a generic type and so on.

So for example if we have a functional interface like:
`
public interface Dummy<IN, OUT> {
OUT apply(Map<String, List<IN>> value)
}
`
To extract inputType **IN** from lambda you need to provide array like `new int[] {0,1,0}`.

**Comment by aljoscha** [comment]
Ah, so it performs nested navigation. If I had a type like
```
public interface Dummy<IN, OUT> { OUT apply(Map<String, Map<List<IN>, String>> value) }
```
it would be `new int[] {0,1,0,1}`?

**Comment by dawidwys** [comment]
In fact `new int[] {0,1,0,0}`, but I think you got the point.

**Comment by dawidwys** [comment]
I think it is ready for a review.

**Comment by aljoscha** [comment]
Thanks, and yes, I see my mistake now.

I'll review, but could you please also have a look, @twalthr? I think you have more experience with this than I do.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 1]
```diff
@@ -422,37 +521,52 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
+					"Indices for output type arguments within lambda not provided");
 				// check for lambda type erasure
 				validateLambdaGenericParameters(exec);
 
 				// parameters must be accessed from behind, since JVM can add additional parameters e.g. when using local variables inside lambda function
-				final int paramLen = exec.getParameterTypes().length - 1;
+				final int paramLen = exec.getParameterTypes().length;
+
+				final Method sam = getSingleAbstractMethod(baseClass);
+				final int baseParametersLen = sam.getParameterTypes().length;
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: What is the difference between `paramLen` and `baseParametersLen`? Is it that the first is the number of generic parameters of the SAM itself and `baseParametersLen` is the number of parameters of the method?

Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Kind of. baseParametersLen is the number of arguments that the method of interface has. So this is the "array" that the indices point to.

paramLen is the number of parameters of the lambda, which can have some additional parameters e.g. from the closure. So we need both of those lengths to index the argument correctly.

Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: I see, thank you!

**Comment by aljoscha** [comment]
Now, let's wait for @twalthr in case he has any comments.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 2]
```diff
@@ -354,49 +479,21 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
-			hasIterable ? 0 : -1,
-			hasCollector ? 0 : -1,
-			inType,
-			functionName,
-			allowMissing);
-	}
-
-	/**
-	 * Returns the unary operator's return type.
+	 * <p><b>NOTE:</b> lambda type indices allows extraction of Type from lambdas. To extract input type <b>IN</b>
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: allow

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java (hunk 1 above)
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Can you add an inline comment to both `paramLen` and `baseParametersLen`?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 3]
```diff
@@ -336,8 +461,8 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
 	public static <T> TypeInformation<T> getPartitionerTypes(Partitioner<T> partitioner, String functionName, boolean allowMissing) {
 		return new TypeExtractor().privateCreateTypeInfo(Partitioner.class, partitioner.getClass(), 0, null, null);
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Can you also update the `Partitioner` to support lambdas?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 4]
```diff
@@ -464,57 +578,52 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
-	public static <IN1, IN2, OUT> TypeInformation<OUT> getBinaryOperatorReturnType(
-		Function function,
-		Class<?> baseClass,
-		boolean hasIterables,
-		boolean hasCollector,
-		TypeInformation<IN1> in1Type,
-		TypeInformation<IN2> in2Type,
-		String functionName,
-		boolean allowMissing) {
+	private static Type extractType(
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Please give this method a more meaningful name (e.g. extractTypeFromLambda) and a Javadoc. You can also move it to `TypeExtractionUtils`.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java [hunk 5]
```diff
@@ -464,57 +578,52 @@ private static void registerFactory(Type t, Class<? extends TypeInfoFactory> fac
...
-			function,
-			baseClass,
-			hasIterables ? 0 : -1,
-			hasCollector ? 0 : -1,
-			in1Type,
-			in2Type,
-			functionName,
-			allowMissing
-		);
+	private static Method getSingleAbstractMethod(Class<?> baseClass) {
```
Comment (twalthr) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java]: Can you document and move this to TypeExtractionUtils?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java [hunk 6]
```diff
@@ -507,12 +505,41 @@ public AllWindowedStream(DataStream<T> input,
...
+
+	private static <IN, OUT> TypeInformation<OUT> getProcessAllWindowFunctionReturnType(
+			ProcessAllWindowFunction<IN, OUT, ?> function,
+			TypeInformation<IN> inType) {
+		return TypeExtractor.getUnaryOperatorReturnType(
+			function,
+			ProcessAllWindowFunction.class,
+			0,
+			1,
+			new int[]{1, 0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: `ProcessAllWindowFunction` is not a SAM. We do not need this parameters.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/CoGroupedStreams.java [hunk 7]
```diff
@@ -231,14 +231,18 @@ protected WithWindow(DataStream<T1> input1,
...
 			TypeInformation<T> resultType = TypeExtractor.getBinaryOperatorReturnType(
-					function,
-					CoGroupFunction.class,
-					true,
-					true,
-					input1.getType(),
-					input2.getType(),
-					"CoGroup",
-					false);
+				function,
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/CoGroupedStreams.java]: We could use `TypeExtractor#getCoGroupReturnTypes()` and save some duplicate code, but I'm also fine with this.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 8]
```diff
@@ -203,9 +203,19 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
-		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(coMapper,
-				CoMapFunction.class, false, true, getType1(), getType2(),
-				Utils.getCallLocationName(), true);
+		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(
+			coMapper,
+			CoMapFunction.class,
+			0,
+			1,
+			2,
+			new int[]{0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 9]
```diff
@@ -227,9 +237,19 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
-		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(coFlatMapper,
-				CoFlatMapFunction.class, false, true, getType1(), getType2(),
-				Utils.getCallLocationName(), true);
+		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(
+			coFlatMapper,
+			CoFlatMapFunction.class,
+			0,
+			1,
+			2,
+			new int[]{0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java [hunk 10]
```diff
@@ -254,9 +274,19 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
...
-		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(coProcessFunction,
-				CoProcessFunction.class, false, true, getType1(), getType2(),
-				Utils.getCallLocationName(), true);
+		TypeInformation<R> outTypeInfo = TypeExtractor.getBinaryOperatorReturnType(
+			coProcessFunction,
+			CoProcessFunction.class,
+			0,
+			1,
+			2,
+			new int[]{0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/ConnectedStreams.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java [hunk 11]
```diff
@@ -574,13 +574,15 @@ public ExecutionConfig getExecutionConfig() {
...
-				false,
-				true,
-				getType(),
-				Utils.getCallLocationName(),
-				true);
+			processFunction,
+			ProcessFunction.class,
+			0,
+			1,
+			new int[]{0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java [hunk 12]
```diff
@@ -221,14 +221,18 @@ protected WithWindow(DataStream<T1> input1,
...
-					input1.getType(),
-					input2.getType(),
-					"Join",
-					false);
+				function,
+				JoinFunction.class,
+				0,
+				1,
+				2,
+				new int[]{0, 0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: Wrong lambda arguments. Should be `(0), (0), ()`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java [hunk 13]
```diff
@@ -300,14 +304,18 @@ protected WithWindow(DataStream<T1> input1,
...
-					input1.getType(),
-					input2.getType(),
-					"Join",
-					false);
+				function,
+				FlatJoinFunction.class,
+				0,
+				1,
+				2,
+				new int[]{0, 0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/JoinedStreams.java]: Wrong lambda arguments. Should be `(0), (1), (2, 0)`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 14]
```diff
@@ -267,13 +267,15 @@ private boolean validateKeyTypeIsHashable(TypeInformation<?> type) {
...
-				false,
-				true,
-				getType(),
-				Utils.getCallLocationName(),
-				true);
+			processFunction,
+			ProcessFunction.class,
+			0,
+			1,
+			new int[]{0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java [hunk 15]
```diff
@@ -981,12 +976,42 @@ public WindowedStream(KeyedStream<T, K> input,
...
+	private static <IN, OUT, KEY> TypeInformation<OUT> getProcessWindowFunctionReturnType(
+			ProcessWindowFunction<IN, OUT, KEY, ?> function,
+			TypeInformation<IN> inType,
+			String functionName) {
+		return TypeExtractor.getUnaryOperatorReturnType(
+			function,
+			ProcessWindowFunction.class,
+			0,
+			1,
+			new int[]{2, 0},
```
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java]: No lambda.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java (hunk 6 above)
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: I think `ProcessAllWindowFunction` is a `FunctionalInterface`. The only abstract method `process`. All other methods have default implementation, so you can pass lambda as `ProcessAllWindowFunction`.

**Comment by rmetzger** [comment]
@dawidwys What's your schedule to address the comments?
This is one of the real blockers for the 1.3.1 release, and I would like to put the first RC this week.

**Comment by dawidwys** [comment]
I will address them today. I am working on last comment with enabling the changes also for Partitioner.

**Comment by rmetzger** [comment]
Cool, thank you!

**Comment by aljoscha** [comment]
@dawidwys I merged to master, now I rebased on `release-1.3` and running tests before merging there as well.

Thanks for fixing this so quickly! 👌

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java (hunk 6 above)
Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: @dawidwys correct me if I'm wrong but as far as I know, a `FunctionalInterface` must be an interface. All `ProcessXXX` functions are regular classes with one abstract methods, but this does not qualify them for lambdas. See also https://stackoverflow.com/questions/24610207/abstract-class-as-functional-interface

Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: You are right! I was not aware of that. Sorry for that. Shall I change it then to `NO_INDEX`?

**Comment by aljoscha** [comment]
@dawidwys Could you please close this PR?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java (hunk 6 above)
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: Ha, it seems I might have merged to fast. What will happen if we leave it as is? Shouldn't the analysis simply fail? We should probably just push a hot fix for this.

Sorry for the inconvenience.

Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: The path that uses the indices will not be reachable anyway, so there are no correctness issues.

Comment (twalthr) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/AllWindowedStream.java]: Yes, it is not a serious problem. But we should change it to be consistent. @dawidwys Can you create a hot fix for it?

**Comment by dawidwys** [comment]
I created a hotfix for the discussed issue: #4089 . I will close this PR then.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java [hunk 16]
```diff
@@ -161,6 +164,77 @@ public static LambdaExecutable checkAndExtractLambda(Function function) throws T
...
+	 */
+	public static Method getSingleAbstractMethod(Class<?> baseClass) {
+		Method sam = null;
+		for (Method method : baseClass.getMethods()) {
+			if (Modifier.isAbstract(method.getModifiers())) {
+				if (sam == null) {
+					sam = method;
+				} else {
+					throw new InvalidTypesException(
+						"Given class: " + baseClass + " is not a FunctionalInterface. It does not have a SAM.");
```
Comment (tedyu) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: This message seems to be inexact: if there is no SAM, sam would be null upon returning from the method.
I suggest changing the message and adding a check (for null sam) prior to returning.

Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractionUtils.java]: @tedyu You are right. I've created a PR with a fix. Could you have a look? #4140

**Pull request closed without merging**